ignite/handlers/ema_handler.py

```diff
@@ -0,0 +1,155 @@
+
+              trainer.run(...)
+
+
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        momentum: float = 0.0002,
+        momentum_warmup: float = 0.0001,
+        interval: int = 1,
+        warmup_iters: int = 100,
+    ) -> None:
+        if not 0 < momentum < 1:
+            raise ValueError(f"Invalid momentum: {momentum}")
+        if not 0 < momentum_warmup < 1:
+            raise ValueError(f"Invalid momentum_warmup: {momentum_warmup}")
+        if not momentum_warmup <= momentum:
+            raise ValueError(
+                f"momentum_warmup should be less than or equal to momentum, but got "
+                f"momentum_warmup: {momentum_warmup} and momentum: {momentum}"
+            )
+        if not (isinstance(interval, int) and interval > 0):
+            raise ValueError(f"Invalid interval: {interval}")
+        if not isinstance(warmup_iters, int) and warmup_iters > 0:
+            raise ValueError(f"Invalid warmup_iters: {warmup_iters}")
+        if not isinstance(model, nn.Module):
+            raise ValueError(
+                f"model should be an instance of nn.Module or its subclasses, but got"
+                f"model: {model.__class__.__name__}"
+            )
+        self.momentum = momentum
+        self.interval = interval
+        self.momentum_warmup = momentum_warmup
+        self.warmup_iters = warmup_iters
+
+        self.unwrapped_model = self._unwrap_model(model)
+        self.ema_model = deepcopy(self.unwrapped_model)
+        self.ema_model.eval()
+
+    @staticmethod
+    def _unwrap_model(model: nn.Module) -> nn.Module:
+        if isinstance(model, (nn.parallel.DataParallel, nn.parallel.DistributedDataParallel)):
+            return model.module
+        else:
+            return model
+
+    def _get_momentum(self, curr_iter: int) -> float:
+        """Get current momentum, `curr_iter` should be 1-based. When `curr_iter = 1`, `momentum =
+        self.momentum_warmup`; when `curr_iter >= self.warmup_iters`, `momentum = self.momentum`"""
+        if curr_iter < 1:
+            raise ValueError(f"curr_iter should be at least 1, but got {curr_iter}.")
+        denominator = max(1, self.warmup_iters - 1)
+        momentum = self.momentum_warmup + (self.momentum - self.momentum_warmup) * (curr_iter - 1) / denominator
+        return min(self.momentum, momentum)
+
+    @torch.no_grad()
+    def _update_weights(self, engine: Engine) -> None:
+        """Update weights of ema model"""
+        momentum = engine.state.ema_momentum
+
+        for ema_v, model_v in zip(self.ema_model.state_dict().values(), self.unwrapped_model.state_dict().values()):
+            ema_v.mul_(1.0 - momentum).add_(model_v.data, alpha=momentum)  # type: ignore
+
+    def _update_ema_momentum(self, engine: Engine) -> None:
+        """Update momentum in engine.state"""
+        curr_iter = engine.state.iteration
+        momentum = self._get_momentum(curr_iter)
```

Comment:
I think `self.momentum` and momentum = `engine.state.ema_momentum` can be misleading. I would also expect `self.momentum` to be current momentum value.

Reply:
`self.momentum` refers to the momentum after warmup, `engine.state.ema_momentum` refers to the current momentum. Do you have any suggestion on re-naming these two variables?

Reply:
Actually, scheduling the momentum is another story I would say. We have currently a draft PR on any parameter scheduling https://github.com/pytorch/ignite/pull/2090 and maybe in future this could be used to set a warm-up or  gradual decay etc...
OK, let's keep it like that for instance..

Reply:
I will leave a TODO marker here. So that we can update it after #2090 is merged